### PR TITLE
[READY] remove inheritance from sklearn Lasso

### PR DESCRIPTION
**Pull request by mathurinm**
split of #19 with only the part concerning the removal of sklearn Lasso inheritance

### CODE REVIEW

File: skglm/estimators.py
Comment (mathurinm) [inline skglm/estimators.py:20]: rm blank line

File: skglm/estimators.py
Comment (mathurinm) [inline skglm/estimators.py:35]: Should we make this private ? Otherwise we need a docstring

File: skglm/estimators.py
Comment (mathurinm) [inline skglm/estimators.py:70]: is this still needed ? it's a bit mysterious so it deserves a comment

File: skglm/estimators.py
Comment (mathurinm) [inline skglm/estimators.py:106]: here I would break line to indicate we start fitting after checking input

File: skglm/estimators.py
Comment (mathurinm) [inline skglm/estimators.py:109]: init should be done after all checks, like checking that X and y shapes match

File: skglm/estimators.py
Comment (mathurinm) [inline skglm/estimators.py:131]: This can be done L97, with an else following https://github.com/scikit-learn-contrib/skglm/pull/54/files#diff-5a55ebbdbb198f456b4549cc660c2b1e6fc838fd05424e8f4b7775239aed32b9R96

File: skglm/estimators.py
Comment (mathurinm) [inline skglm/estimators.py:139]: Make the error message explicit: expected..., got ...

File: skglm/estimators.py
Comment (mathurinm) [inline skglm/estimators.py:169]: Logistic is for classif, I believe we will never this block ?

File: skglm/solvers/cd_solver.py
Comment (mathurinm) [inline skglm/solvers/cd_solver.py:348]: this can break existing code. Is it completely needed ? Can you get it from outside by multiplying `len(obj_out)` by 10 ?

File: skglm/estimators.py
Comment (Klopfe) [reply skglm/estimators.py:35]: I have added docstring for now. I can also be private function eventually

Comment (mathurinm) [reply skglm/estimators.py:35]: I think it should be private. Do you envision a case where a user would need it?

Comment (Klopfe) [reply skglm/estimators.py]: Not really, I changed it to a private function :)

File: skglm/estimators.py
Comment (Badr-MOUFAD) [inline skglm/estimators.py:93]: these cases are special cases (rep WeighedLasso). I think they should be treated inside the model rather than a function with if/else statements.

File: skglm/estimators.py
Comment (Badr-MOUFAD) [inline skglm/estimators.py:110]: same remark as ``QuadraticSVC`` in L93

File: skglm/estimators.py
Comment (mathurinm) [inline skglm/estimators.py:103]: this fails if coef_ is None

File: skglm/estimators.py
Comment (mathurinm) [inline skglm/estimators.py:124]: add a comment: n_iter for us means number of subproblems (for sklearn it's number of CD epochs, not the same)

File: skglm/estimators.py
Comment (mathurinm) [inline skglm/estimators.py:141]: harmonize style with below: break after (

**Review by Badr-MOUFAD — COMMENTED** [review]
The PR is much simpler to review now. I think we are on the right way to getting rid of sklearn inheritance, though I have the following remark:
- In our estimators, we have a fit and path methods that don't seem to use the same things (``cd_solver`` for path, and ``_glm_fit`` for fit)

and code review.

File: skglm/estimators.py
Comment (QB3) [reply skglm/estimators.py:33]: I am on it!

Comment (QB3) [reply skglm/estimators.py]: this should be OK now

File: skglm/estimators.py
Comment (QB3) [reply skglm/estimators.py:142]: I am not sure that we will be able to put in the datafit for the SVM.
The problem is that the SVM is solved in the dual, if we really want to take care of in this PR, I think the cleanest way would be to add a `solved_in_dual` parameter

Comment (mathurinm) [reply skglm/estimators.py:142]: You can't treat that in the model because this function is called by GeneralizedLinearEstimator

File: skglm/estimators.py
Comment (mathurinm) [reply skglm/estimators.py]: I am not sure this is relevant. How much duplicated code do you have now ?

File: skglm/estimators.py
Comment (QB3) [reply skglm/estimators.py:171]: it is already in the doc of estimators: "max_iter : int, optional The maximum number of iterations (subproblem definitions).", do you want to make a recall here?

**Comment by mathurinm** [comment]
Note that currently I don't think we can use GeneralizeLinearEstimator with a group penalty, has it will not call `group_bcd_solver`. I wonder how the tests don't break

**Comment by QB3** [comment]
> Note that currently I don't think we can use GeneralizeLinearEstimator with a group penalty, has it will not call group_bcd_solver. I wonder how the tests don't break

so running `test_group.py` should fail

File: skglm/tests/test_estimators.py
Comment (QB3) [inline skglm/tests/test_estimators.py:230]: is there a reason not to test the warm start for all the estimators?

File: skglm/estimators.py
Comment (QB3) [reply skglm/estimators.py]: I am not sure to understand what you mean here

File: skglm/estimators.py
Comment (QB3) [reply skglm/estimators.py]: should we merge it back then?

**Comment by QB3** [comment]
> Note that currently I don't think we can use GeneralizeLinearEstimator with a group penalty, has it will not call group_bcd_solver. I wonder how the tests don't break

`GeneralizeLinearEstimator` is currently not tested for group estimators, hence it makes sense that tests do not fail.
In `test_group.py` we directly call `group_bcd_solver`.

**Comment by mathurinm** [comment]
OK, great. When we add the `Solver` class we should test all solvers with the `GLE` class

File: skglm/tests/test_estimators.py
Comment (mathurinm) [reply skglm/tests/test_estimators.py:230]: I just wanted the test to be fast. I was thinking if it worked for one estimator it would work for all of them

File: skglm/estimators.py
Comment (mathurinm) [reply skglm/estimators.py:96]: Nope you are right, I forgot it was there

File: skglm/estimators.py
Comment (mathurinm) [reply skglm/estimators.py:79]: we can have a penalty.check_compatible(X, y)  but that's more work for the user when they define penalties.

File: skglm/estimators.py
Comment (mathurinm) [inline skglm/estimators.py:64]: This is not needed in this function, it's reminiscent from glm_classif. It could be confusing. I would merge the two functions back

File: skglm/estimators.py
Comment (mathurinm) [inline skglm/estimators.py:167]: @Klopfe why is p0 equal to X_.shape here ? Avoid writing such code without a TODO comment, it can very well go unnoticed, then be merged and it would take a long time to find it back

Comment (Klopfe) [reply skglm/estimators.py:167]: I don't think that the "why" question is necessary. I did not leave a mistake on purpose. I was debugging the SVC. Now, you're right I should have been more cautious. I will pay attention. next time.

Comment (mathurinm) [reply skglm/estimators.py:167]: I was asking if if was for debugging purposes or if there was a reason it was still there. Sorry if the tone was not friendly, it was not my intention.

Comment (Klopfe) [reply skglm/estimators.py:167]: I misunderstood then. It should have been back to p0=10 and it could have been very long to find out. You're right

File: skglm/estimators.py
Comment (QB3) [reply skglm/estimators.py:28]: done!

File: skglm/estimators.py
Comment (QB3) [reply skglm/estimators.py]: I went back to `p0=model.p0,` should we put a different value for  the SVC?

File: skglm/estimators.py
Comment (QB3) [inline skglm/estimators.py:143]: the `ws_strategy` was commented, should we add it back? or do not let the choice to the user?

File: skglm/tests/test_estimators.py
Comment (QB3) [reply skglm/tests/test_estimators.py]: OK, I just added all the estimators to the test!

File: skglm/estimators.py
Comment (QB3) [reply skglm/estimators.py:33]: was it taken from here ? https://github.com/scikit-learn/scikit-learn/blob/5bd81234e6e6501ddcddbfdfdc80b90a1302af55/sklearn/svm/_base.py#L1177

**Comment by QB3** [comment]
It would be nice to add a comment to the `test_generic_get_params` function in `test_estimators.py`

**Comment by QB3** [comment]
Apart from the few comments left, LGTM

File: skglm/estimators.py
Comment (Klopfe) [inline skglm/estimators.py:139]: I think we can leave the p0 as it is for the SVC, I will investigate the slow tests to see if I can find why it happens.

File: skglm/estimators.py
Comment (Klopfe) [reply skglm/estimators.py:144]: Maybe we should just do for convex penalties = subdiff and for non-convex: fixed_point until we have time to do more benchmarks. Wdyt?

File: skglm/estimators.py
Comment (Klopfe) [inline skglm/estimators.py:164]: It looks neat. Well done 👍

File: skglm/estimators.py
Comment (Klopfe) [inline skglm/estimators.py:402]: since _glm_fit is merged back, we should just return _glm_fit once here.

File: skglm/estimators.py
Comment (Klopfe) [inline skglm/estimators.py:1236]: I will add as a TODO for another PR to add a predict_proba for LinearSVC, probably just taking some code from sklearn.

File: skglm/estimators.py
Comment (Klopfe) [reply skglm/estimators.py:402]: I fixed it

File: skglm/estimators.py
Comment (mathurinm) [inline skglm/estimators.py:90]: cosmetic: when two if blocks are unrelated (not like an if elif), it often helps to put a blank line between them

**Comment by mathurinm** [comment]
LGTM

File: skglm/estimators.py
Comment (Klopfe) [reply skglm/estimators.py:144]: Leave as it is for now and we will see in the future :)

File: skglm/estimators.py
Comment (Badr-MOUFAD) [inline skglm/estimators.py:1491]: Is there a reason why we don't call ``_glm_fit`` here?

File: skglm/estimators.py
Comment (Badr-MOUFAD) [inline skglm/estimators.py:253]: We have two big chunks of commented code in ``GeneralizedLinearEstimator``, What are we planning to do with them? (remove, keep for future refs, )

**Review by Badr-MOUFAD — COMMENTED** [review]
Congrats on this amazing job! I think we are almost done with inheritance from scikit-learn estimators.

I have two comments before we merge.

File: skglm/estimators.py
Comment (Klopfe) [reply skglm/estimators.py:1491]: Here, we want to compute the path for several alphas. Hence, calling `multitask_bcd_solver_path` as it is done in the Lasso case with `cd_solver_path`. So I'm guessing it is all right

File: skglm/estimators.py
Comment (mathurinm) [reply skglm/estimators.py:253]: they can be removed

Comment (Klopfe) [reply skglm/estimators.py:253]: To my opinion, the code in fit can be removed as it can now be found almost completely in `_glm_fit`. I don't really see reasons to keep it but maybe the others will think otherwise.

File: skglm/estimators.py
Comment (Badr-MOUFAD) [reply skglm/estimators.py:1491]: What about the fit method, why don't call ``_glm_fit``? In fact, we handle the case of MLT (c.f. L135).

Comment (Klopfe) [reply skglm/estimators.py:1491]: You're right but we don't handle multiple penalty values in the `_glm_fit`.

**Review by Badr-MOUFAD — APPROVED** [review]

**Comment by Badr-MOUFAD** [comment]
LGTM when it's green!

**Comment by Klopfe** [comment]
Well done everyone

**Pull request merged**
